[ENG-9487][docs] add docs for custom builds (#23880)
expo/expo: docs/pages/eas-build-custom-builds/custom-build-config.mdx

import { BoxLink } from '~/ui/components/BoxLink';
import { GithubIcon } from '@expo/styleguide-icons';

> **Warning** This functionality is in preview, under active development, and it can be subject to changes.

{/* TODO: link to the blog post */}

> > **Warning** Other, current limitations of custom builds are highlighted in the launch blog post and include: not adapted pricing, limitations regarding uploading artifacts, lack of reusable steps for caching dependencies, no support for classic updates, and not ready GitHub integration for custom builds.

Reviewer: ```suggestion
> **Warning** Current limitations of custom builds are highlighted in the [changelog](https://expo.dev/changelog/2023/08-10-custom-build-jobs) and include not adapted pricing, limitations regarding uploading artifacts, lack of reusable steps for caching dependencies, no support for classic updates, and not ready GitHub integration for custom builds.
```

Added the changelog post here since the blog post will also be copy of it but the changelog post will be ready before the blog post.

  "build": {
    /* @hide ... */ /* @end */
    "test": {
      "config": "test.yml",
      "withoutCredentials": true
    },
}
```

> We can also use the `withoutCredentials` option to skip the credentials setup on the CLI side as they are not needed for running tests.

Reviewer: ```suggestion
You can also use the `withoutCredentials` option to skip the credentials setup on the CLI side, as they are not needed for running tests.
```

[Let's phrase this sentence in second person voice](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md#write-in-the-second-person)

> If you wish to use different, separate configs for each platform, you can create separate YAML config files for iOS and Android. For example:

Reviewer: ```suggestion
If you wish to use separate configs for each platform, you can create separate YAML config files for Android and iOS. For example:
```

Phrased as per [Referencing platforms](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md#referencing-android-ios-and-web).

## Summary

- Build configs are customizable by adding steps you want to run on the EAS service as an extension to your existing regular builds.
- You can configure the credentials for custom builds the same way as for regular builds. Especially if you want to run a custom build that doesn't require credentials, you can use the `withoutCredentials` option as shown above. See [App credentials docs](/app-signing/app-credentials/) for more information.
- If your regular builds have [EAS secrets](/build-reference/variables/#adding-secrets-with-eas-cli), they are also available for custom builds.

> Also, check out the **example repository** for more detailed examples:

Reviewer: ```suggestion
Check out the **example repository** for more detailed examples:
```
